File: Leetcode/35-Search-Insert-Position/SearchInsertPosition.java
public class SearchInsertPosition {

    public int searchInsert(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return 0;
        }

        return searchInsert(nums, target, 0, nums.length-1);
    }

    private int searchInsert(int[] nums, int target, int start, int end) {
        int mid = start + (end - start)/2;

        if (target == nums[mid]) {
            return mid;
        } else if (target < nums[mid]) {
            return start < mid? searchInsert(nums, target, start, mid-1):start;
        } else {
            return end > mid? searchInsert(nums, target, mid+1, end):(end+1);
        }
    }
}

File: Leetcode/35-Search-Insert-Position/SearchInsertPosition.cpp
class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        if (nums.empty()) {
            return 0;
        }
        int start = 0;
        int end = nums.size()-1;
        while (start < end) {
            int mid = start + (end - start)/2;
            if (nums[mid] == target) {
                return mid;
            }
            if (nums[mid] > target) {
                end = mid -1;
            } else {
                start = mid +1;
            }
        }
        // start == end
        if (nums[start] < target) {
            return start+1;
        }
        return start;
    }
};
